### ISSUE TITLE
Typescript support is broken

### ISSUE THREAD

**Issue by Dan503** [gulpjs/gulp#2623]
### What were you expecting to happen?

When I follow the steps outlined in this stack-overflow post, it should allow us to write our gulp files in pure Typescript:

https://stackoverflow.com/a/49849088/1611058

These were steps that used to work back in 2020 but don't work anymore.

### What actually happened?

As sydd's comment says:

> I get `SyntaxError: Cannot use import statement outside a module` if the project's type is commonjs and `TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts"` if the project's type is module

### Please give us a sample of your gulpfile

```js
// gulpfile.ts

import gulp from 'gulp'

gulp.task('default', function (done) {
    console.log('test');
    done()
});
```

### Terminal output / screenshots

![image](https://user-images.githubusercontent.com/10610368/130347865-56e8b664-0173-425d-afd5-b9972e4c59d6.png)

![image](https://user-images.githubusercontent.com/10610368/130347914-8bf56f5b-4b62-4149-a603-2846485964a4.png)

### Please provide the following information:
* OS & version: Windows 10 - 20H2
* node version (run `node -v`): v12.18.2
* npm version (run `npm -v`): 6.14.5
* gulp version (run `gulp -v`):  CLI version: 2.3.0; Local version: 4.0.2

**Comment by phated** [comment]
You are using esm, so this is a duplicate of https://github.com/gulpjs/rechoir/issues/43 - it requires an insane amount of work that we probably won't get to for awhile. You can switch back to CJS.

**Comment by Dan503** [comment]
@phated I got those errors when I tried to add Gulp to a [Vite](https://vitejs.dev/) project.

Do you know a way to ensure that Gulp uses CJS while Vite can still use ESM?

**Comment by phated** [comment]
I don't have generic advice, sorry. Would have to take it case-by-case and I don't have the bandwidth right now.

**Comment by Dan503** [comment]
> You can switch back to CJS.

Basically I'm asking, is it possible to use both ESM and Common JS in the same project?

ESM gets used by some stuff while Gulp uses Common JS.

**Comment by Dan503** [comment]
I've asked on Stack Overflow.

https://stackoverflow.com/q/68899837/1611058

I hope it's possible...

**Comment by phated** [comment]
Yes, you need to remove the `module` field in package.json and then you need to use `.mjs` on all your ESM files. However, you might need 2 different tsconfig.json files. I don't know - I hate typescript...

**Comment by Dan503** [comment]
I've investigated the issue a bit more.

This is an issue with [ts-node](https://github.com/TypeStrong/ts-node), not Gulp.

It is highly related to this issue:
https://github.com/TypeStrong/ts-node/issues/922

I created a reduced test case and was able to determine that *something* in the tsconfig file causes the error in ts-node.

When I added a tsconfig file to the project, it caused the error to happen.
When I removed the tsconfig file, gulp worked perfectly.

Not an issue for Gulp maintainers but still an issue in general since tsconfig files are important.
At least I've managed to narrow down the problem a bit.
